Author guide updates

joe-lawrence opened the pull request:
A few minor updates to the author guide, including scrubbing kpatch.ko and the old stop_machine model from the doc:  link to upstream kernel API documentation where applicable.

joe-lawrence commented:
note:  I just noticed some kpatch.ko specific info in the main README.md file, like the `KPATCH_FORCE_UNSAFE` macro... I can exorcise that stuff in another PR.

joe-lawrence commented:
Added sections on constprop and sibling calls.

jpoimboe commented on doc/patch-author-guide.md:
```diff
@@ -261,58 +224,76 @@ A shadow PID variable is allocated in `do_fork()`: it is associated with the
 +
-+		newpid = kpatch_shadow_alloc(p, "newpid", sizeof(*newpid),
-+					     GFP_KERNEL);
++		newpid = klp_shadow_get_or_alloc(p, 0, sizeof(*newpid), GFP_KERNEL,
```
Maybe have `#define KPATCH_SHADOW_NEWPID 0` somewhere (or enum) and use that instead?  Perhaps along with a comment that the value needs to be globally unique.

joe-lawrence replied:
ok

jpoimboe commented on doc/patch-author-guide.md:
```diff
@@ -324,9 +305,9 @@ structure can be used by patched code to handle both new (post-patch) and
-required to allocate/free a shadow variable called "reqs_active_v2" whenever a
-new `struct kioctx` is created/released. No values are ever assigned to the
-shadow variable.)
+required to allocate/free a shadow variable with enum ID SV_REQS_ACTIVE_V2
```
I had to do a double take at the SV_ prefix, maybe give it a KPATCH_SHADOW prefix instead?  Also it might be clearer to define it somewhere in the code.

joe-lawrence replied:
KPATCH_SHADOW_ prefix is kinda long.  Is something like ID_, KLP_SV, or KPS_ too cryptic?

jpoimboe replied:
If it makes it more readable, why not go with the longer version?  The extra characters are free.  KLP_SV and KPS_ are cryptic, and ID_ doesn't make clear that it's a kpatch-specific id.

jpoimboe commented on doc/patch-author-guide.md:
```diff
@@ -386,7 +367,8 @@ the spinlock:
  
  	spin_lock_init(&sta->lock);
-+	ps_lock = kpatch_shadow_alloc(sta, "ps_lock", sizeof(*ps_lock), gfp);
++	ps_lock = klp_shadow_alloc(sta, PS_LOCK, sizeof(*ps_lock), gfp,
```
Similar comment here for PS_LOCK

joe-lawrence commented on doc/patch-author-guide.md:
```diff
@@ -193,66 +197,25 @@ static void kpatch_post_unpatch_tcp_send_challenge_ack(patch_object *obj)
-is held to block the patching.
+Also note the pre-patch callback return code will be ignored by the kernel's
+module notifier, so it does not affect the target module or livepatch module
+status.
```
Hmm, actually the original text is completely wrong.  The upstream documentation states:
>If a pre-patch callback returns failure, the kernel's module loader will:
>  - Refuse to load a livepatch, if the livepatch is loaded after targeted code.
>    or:
>  - Refuse to load a module, if the livepatch was already successfully loaded.

I'll update this accordingly.

joe-lawrence commented on doc/patch-author-guide.md:
```diff
@@ -261,58 +224,76 @@ A shadow PID variable is allocated in `do_fork()`: it is associated with the
 sized accordingly, and allocated as per `GFP_KERNEL` flag rules.
 
-`kpatch_shadow_alloc` returns a pointer to the shadow variable, so we can
+`klp_shadow_get_or_alloc()` returns a pointer to the shadow variable, so we can
```
FWIW, I had grabbed this from a local test/integration/fedora-30/shadow-newpid.patch, not realizing that it was never pushed up to the repository.  I think either call should work.  klp_shadow_get_or_alloc() is probably more straightforward.

jpoimboe replied:
How would klp_shadow_get_or_alloc() be more straightforward here?  In do_fork() there shouldn't be an existing shadow struct for it, right?

joe-lawrence replied:
D'oh.. copy/paste error.  I meant to say klp_shadow_alloc() for that reason.

joe-lawrence replied:
Actually.... this bugged me as I found the same klp_shadow_get_or_alloc() usage in the rhel-7.x integration tests.  It traces back to the centos-7 tests in which @sm00th modified the test in 56be80720aee ("integration/centos-7: fix shadow-newpid.patch warnings").  From the commit message:
> When quickly loading/unloading this patch multiple times it is possible to hit "Duplicate shadow variable" warnings since the patch doesn't have any cleanup hooks on unload.
Switch to klp_shadow_get_or_alloc to ignore these.

I can provide the original version of the patch in the author guide using klp_shadow_alloc() and mention that it was adjusted for testing purposes or some other disclaimer.

jpoimboe replied:
Ah, that makes sense now.  I guess the correct fix would be to use klp_shadow_free_all() in a post-unload callback?  If so it might be worth adding that to the real integration patches as well for proper test coverage.

joe-lawrence replied:
I would say that is the correct solution for a real kpatch, however I usually try hard not to mix features when writing test cases.  In this case, I think we could call klp_shadow_free_all() from a module_exit function to achieve the same result.

joe-lawrence commented on doc/patch-author-guide.md:
```diff
@@ -597,3 +579,126 @@ For example, a `pr_warn_once()` can be replaced with:
+changes to all of `__tcp_mtu_to_mss()` callers (ie, it was inlined as
+requested).  In this case, a simple workaround is to provide a new,
+non-inlined copy of `__tcp_mtu_to_mss()` and then update all its callers
+to use it.
```
Yeah I think that would work, too.

joe-lawrence commented on doc/patch-author-guide.md:
```diff
@@ -597,3 +579,126 @@ For example, a `pr_warn_once()` can be replaced with:
+-------------
+
+If function A() calls function B() and they share a common function
+signature, GCC may deem them "sibling calls" and apply a tail call
```
Hmm, I suppose that is true, I haven't looked into the implementation of how they implement it.  As long as the stack expectations line up when B() is called, I can see how the signatures don't need to match exactly.  I can reword this to state that the compiler may optimize A() calling B() at the end of the function, skipping some calling conventions to save overhead.

jpoimboe replied:
The signatures don't need to match because for sibling calls, A() restores the stack to its callee's state, and then sets up B()'s args appropriately before jumping to B().

joe-lawrence commented on doc/patch-author-guide.md:
```diff
@@ -597,3 +579,126 @@ For example, a `pr_warn_once()` can be replaced with:
+If function A() calls function B() and they share a common function
+signature, GCC may deem them "sibling calls" and apply a tail call
+optimization in which the call to B() avoids the overhead of an ordinary
+function call.  This may be an issue for kpatches which only A() or B():
```
Ah, yes "patch A() or B()".  I'll add a bit about ppc64le and "external" calls.

joe-lawrence commented:
Another round of revisions.. I redid the shadow variable diff hunks to reflect the new API and @jpoimboe 's suggestions.  A few other nits found here and there (kpatch.ko references keep popping up the more I read the doc.)

jpoimboe commented on doc/patch-author-guide.md:
````diff
@@ -680,3 +680,24 @@ if (static_key_true(&key1))
+  138: 0000000000000000    568 FUNC    LOCAL  DEFAULT       50 timekeeping_update.constprop.9
+```
+The current workaround for this issue to add `__attribute__((noclone))`
+to offending functions.
````
I think this issue was fixed in #1043.  Should this commit be dropped?

joe-lawrence replied:
Yeah, I think it can go

jpoimboe commented on doc/patch-author-guide.md:
```diff
@@ -701,3 +701,23 @@ correleating constprop subfunctions created by the optimization:
+excellent example and description of this problem with annotated disassembly.
+
+Adding `__attribute__((optimize("-fno-optimize-sibling-calls")))` instructs
+GCC to turn off the optimization for the given function.
```
Reading this, it's not clear how the user should know they've hit this issue in the first place.  It should probably mention the error message:
```
			ERROR("Found an unsupported sibling call at %s()+0x%lx.  Add __attribute__((optimize(\"-fno-optimize-sibling-calls\"))) to %s() definition.",
```

joe-lawrence replied:
Good point, I'll add it to the beginning of the section.

joe-lawrence commented:
Addressed @jpoimboe 's comments on the previous version and added a few more static key examples.  In between revisions, there have been a few new issues opened concerning the guide (callback best practices, etc.) that we can address with a follow up PR.  This one has been neglected long enough :)
